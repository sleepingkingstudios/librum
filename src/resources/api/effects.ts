import {
  camelCase,
  kebabCase,
} from 'lodash';

import type { DisplayAlertProps } from '@alerts';
import type { AlertDirective } from '@api/effects/display-alerts';
import {
  pastTense,
  singularize,
} from '@utils/inflector';

const generateContext = ({
  resourceName,
  scope,
}: {
  resourceName: string,
  scope?: string,
}): string => {
  if (scope && scope.length > 0) {
    return `resources:${camelCase(scope)}:${camelCase(resourceName)}:request`;
  }

  return `resources:${camelCase(resourceName)}:request`;
}

export const generateAlerts = ({
  action,
  member,
  query,
  resourceName,
  singularName,
  scope,
}: {
  action: string,
  member: boolean,
  query: boolean,
  resourceName: string,
  singularName?: string,
  scope?: string,
}): AlertDirective[] => {
  const context = generateContext({ resourceName, scope });
  const friendlyName =
    kebabCase(
      member ?
        (singularName || singularize(resourceName)) :
        resourceName
    )
    .replace('-', ' ');
  const failureAlert: DisplayAlertProps = {
    context,
    message: `Unable to ${action} ${friendlyName}`,
    type: 'failure',
  };

  if (query) {
    return [
      {
        display: failureAlert,
        status: 'errored',
      },
      {
        display: failureAlert,
        status: 'failure',
      },
    ];
  }

  const successAlert: DisplayAlertProps = {
    context,
    message: `Successfully ${pastTense(action)} ${friendlyName}`,
    type: 'success',
  };

  return [
    {
      display: failureAlert,
      status: 'errored',
    },
    {
      display: failureAlert,
      status: 'failure',
    },
    {
      display: successAlert,
      status: 'success',
    },
  ];
};
