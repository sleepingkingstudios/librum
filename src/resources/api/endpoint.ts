import type {
  EndpointDefinition as NativeEndpointDefinition,
} from '@reduxjs/toolkit/query/react';

import { api } from '@api';
import type {
  ApiSuccess,
  UseMutation,
  UseQuery,
} from '@api';
import { upperCamelCase } from '@utils/text';
import type {
  ResourceApiEndpoint,
  ResourceConfiguration,
} from '../types';
import type {
  ResourceMutationParams,
  ResourceQueryParams,
} from './types';
import {
  generateEndpointName,
  generateResourceUrl,
} from './utils';

type BuildEndpointProps = {
  action: string,
  endpointName: string,
} & InjectEndpointProps;

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type BuilderQueryFunction = <Response = unknown, Param = unknown>({
  query,
}: {
  query: InnerQueryFunction,
// eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => NativeEndpointDefinition<any, any, any, any, string>;

type InjectEndpointProps = {
  endpoint: ResourceApiEndpoint,
} & ResourceConfiguration;

type InnerQueryFunction = (arg?: unknown) => {
  body?: Record<string, unknown>,
  method: string,
  params?: Record<string, string>,
  url: string,
};

export type Builder = {
  mutation: BuilderQueryFunction,
  query: BuilderQueryFunction,
};

const buildMutationEndpoint = ({
  action,
  baseUrl,
  endpoint,
  endpointName,
  resourceName,
  scope,
}: BuildEndpointProps) => {
  const {
    member,
    method,
    url,
  } = endpoint;

  return (builder: Builder) => ({
    [endpointName]: builder.mutation<ApiSuccess<Record<string, unknown>>, unknown>({
      query: (arg?: ResourceMutationParams) => {
        const { body, params, wildcards } = arg || {};

        const resourceUrl = generateResourceUrl({
          action,
          baseUrl,
          member,
          resourceName,
          scope,
          url,
          wildcards,
        });

        return {
          ...(body ? { body }: {}),
          ...(params ? { params }: {}),
          method: method.toUpperCase(),
          url: resourceUrl,
        };
      },
    }),
  });
};

const buildQueryEndpoint = ({
  action,
  baseUrl,
  endpoint,
  endpointName,
  resourceName,
  scope,
}: BuildEndpointProps) => {
  const {
    member,
    method,
    url,
  } = endpoint;

  return (builder: Builder) => ({
    [endpointName]: builder.query<ApiSuccess<Record<string, unknown>>, unknown>({
      query: (arg?: ResourceQueryParams) => {
        const { params, wildcards } = arg || {};

        const resourceUrl = generateResourceUrl({
          action,
          baseUrl,
          member,
          resourceName,
          scope,
          url,
          wildcards,
        });

        return {
          ...(params ? { params }: {}),
          method: method.toUpperCase(),
          url: resourceUrl,
        };
      },
    }),
  });
};

const buildEndpoint = ({
  action,
  baseUrl,
  endpoint,
  endpointName,
  resourceName,
  scope,
}: BuildEndpointProps) => {
  const { method } = endpoint;
  const builder = method.toLowerCase() === 'get' ?
    buildQueryEndpoint :
    buildMutationEndpoint;

  return builder({
    action,
    baseUrl,
    endpoint,
    endpointName,
    resourceName,
    scope,
  });
};

export const injectEndpoint = ({
  baseUrl,
  endpoint,
  resourceName,
  singularName,
  scope,
}: InjectEndpointProps): Record<string, UseQuery | UseMutation> => {
  const {
    action,
    member,
    method,
  } = endpoint;
  const endpointName = generateEndpointName({
    action,
    member,
    resourceName,
    singularName,
    scope,
  });
  const endpoints = buildEndpoint({
    action,
    baseUrl,
    endpoint,
    endpointName,
    resourceName,
    scope,
  });
  const injected =
    api.injectEndpoints({ endpoints }) as Record<string, unknown>;
  const key = `use${upperCamelCase(action)}Resource${member ? '' : 's'}`;
  let useEndpoint;

  if (method.toLowerCase() === 'get') {
    useEndpoint = injected[`use${endpointName}Query`] as UseQuery;
  } else {
    useEndpoint = injected[`use${endpointName}Mutation`] as UseMutation;
  }

  return { [key]: useEndpoint };
};
