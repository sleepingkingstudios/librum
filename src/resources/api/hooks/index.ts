export { useResourceQuery } from './use-resource-query';
export type { UseResourceQueryOptions } from './use-resource-query';
