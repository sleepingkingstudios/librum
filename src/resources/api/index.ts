export { useResourceQuery } from './use-resource-query';
export type { UseResourceQueryOptions } from './use-resource-query';
export {
  generateAlerts,
  generateUrl,
} from './utils';
