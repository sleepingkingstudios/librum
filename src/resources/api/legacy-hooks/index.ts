export { useResourceQuery } from './use-resource-query';
