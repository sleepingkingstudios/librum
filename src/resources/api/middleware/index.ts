export { handleNotFoundMiddleware } from './handle-not-found';
