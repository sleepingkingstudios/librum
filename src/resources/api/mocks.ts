import type { UseQuery } from '@api';
import { defaultResult } from '@api/test-helpers';
import type {
  ResourceApiEndpointConfiguration,
  ResourceConfiguration,
} from '../types';
import type { ResourceApiHooks } from './types';

const useIndexResources: jest.MockedFunction<UseQuery> =
  jest.fn(() => defaultResult);

type GenerateResourcesApiProps = {
  endpoints?: ResourceApiEndpointConfiguration,
} & ResourceConfiguration;

type GenerateResourcesApi =
  (options: GenerateResourcesApiProps) => ResourceApiHooks;

export const generateResourcesApi: jest.MockedFunction<GenerateResourcesApi> =
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  jest.fn((...args) => ({
    useIndexResources,
  }));
