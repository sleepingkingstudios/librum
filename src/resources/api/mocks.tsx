import { responseWithStatus } from '@api';
import type {
  Refetch,
  Response,
} from '@api';

type UseResourceQuery = (props: unknown) => [Response, Refetch];

export { generateAlerts } from './utils';

export const useResourceQuery: jest.MockedFunction<UseResourceQuery> = jest.fn(
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  (props) => [responseWithStatus({ status: 'uninitialized' }), jest.fn()]
);
