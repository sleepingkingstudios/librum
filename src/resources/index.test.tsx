import * as React from 'react';

import '@testing-library/jest-dom';
import { Routes } from 'react-router-dom';

import { generateResource } from './index';
import { responseWithData } from '@api';
import type { DataTableData } from '@components/data-table';
import { render } from '@test-helpers/rendering';
import { useResourceQuery } from './api';
import type {
  DataBlockData,
  ResourceProps,
} from './types';

// eslint-disable-next-line @typescript-eslint/no-unsafe-return
jest.mock('@components/page', () => require('@components/page/mocks'));
jest.mock('./api');

const mockUseResourceQuery =
  useResourceQuery as jest.MockedFunction<typeof useResourceQuery>;

const Block = ({ data }: { data: DataBlockData }): JSX.Element => {
  const book = 'rareBook' in data ? data.rareBook : { name: 'Unknown Book' };

  return (
    <span className="mock-data-table">
      { `Showing book "${book.name.toString()}"` }
    </span>
  );
};
const Table = ({ data }: { data: DataTableData }): JSX.Element => {
  const books = 'rareBooks' in data ? data.rareBooks : [];

  return (
    <span className="mock-data-table">
      { `There are ${books.length} books!` }
    </span>
  );
};

describe('Resources generateResource()', () => {
  const shouldGenerateTheIndexRoute = ({
    at = '/rare-books',
    resource,
  }: {
    at?: string,
    resource: ResourceProps,
  }): void => {
    describe('should generate the index route', () => {
      const indexData: DataTableData = {
        rareBooks: [
          {
            name: 'Gideon the Ninth',
          },
          {
            name: 'Harrow the Ninth',
          },
          {
            name: 'Nona the Ninth',
          },
        ],
      };
      const refetch = jest.fn();
      const response = responseWithData({ data: indexData });
      const { routes } = generateResource(resource);

      beforeEach(() => {
        mockUseResourceQuery.mockImplementation(() => [response, refetch]);
      });

      it('should render the content', () => {
        const { getByText } = render(
          <Routes>{ routes() }</Routes>,
          {
            initialEntries: [at],
            router: true,
            store: true,
          },
        );

        expect(getByText('There are 3 books!')).toBeVisible();
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <Routes>{ routes() }</Routes>,
          {
            initialEntries: [at],
            router: true,
            store: true,
          },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  };
  const shouldGenerateTheShowRoute = ({
    at = '/rare-books/gideon-9',
    resource,
  }: {
    at?: string,
    resource: ResourceProps,
  }): void => {
    describe('should generate the show route', () => {
      const showData: DataBlockData = {
        rareBook: {
          name: 'Gideon the Ninth',
          slug: 'gideon-9',
          author: 'Tammsyn Muir',
        },
      };
      const refetch = jest.fn();
      const response = responseWithData({ data: showData });
      const { routes } = generateResource(resource);

      beforeEach(() => {
        mockUseResourceQuery.mockImplementation(() => [response, refetch]);
      });

      it('should render the content', () => {
        const { getByText } = render(
          <Routes>{ routes() }</Routes>,
          {
            initialEntries: [at],
            router: true,
            store: true,
          },
        );

        expect(getByText('Showing book "Gideon the Ninth"')).toBeVisible();
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <Routes>{ routes() }</Routes>,
          {
            initialEntries: [at],
            router: true,
            store: true,
          },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  };
  const shouldGenerateTheStandardRoutes = ({
    at = '/rare-books',
    resource,
  }: {
    at?: string,
    resource: ResourceProps,
  }): void => {
    shouldGenerateTheIndexRoute({ at, resource });

    shouldGenerateTheShowRoute({ at: `${at}/gideon-9`, resource });
  };

  const resourceName = 'rareBooks';
  const resource: ResourceProps = { Block, Table, resourceName };

  it('should be a function', () => {
    expect(typeof generateResource).toBe('function');
  });

  it('should generate the resource Pages', () => {
    const { Pages } = generateResource(resource);
    const expected = ['IndexPage', 'ShowPage'];

    expect(Object.keys(Pages)).toEqual(expected);
  });

  shouldGenerateTheStandardRoutes({ resource });

  describe('with scope: value', () => {
    const scope = 'lendingLibrary';
    const resource: ResourceProps = { Block, Table, resourceName, scope };

    shouldGenerateTheStandardRoutes({
      at: '/lending-library/rare-books',
      resource,
    });
  });

  describe('with extra pages', () => {
    const endpoints = {
      publish: {
        member: true,
        method: 'PUT',
      },
      published: {
        member: false,
      },
    };
    const pages = {
      publish: {
        contents: (<span>Publish Page</span>),
        member: true,
      },
      published: {
        contents: (<span>Published Page</span>),
      },
    };
    const resource = {
      Block,
      Table,
      endpoints,
      pages,
      resourceName,
    };

    it('should generate the resource Pages', () => {
      const { Pages } = generateResource(resource);
      const expected = [
        'IndexPage',
        'ShowPage',
        'PublishPage',
        'PublishedPage',
      ];

      expect(Object.keys(Pages)).toEqual(expected);
    });

    shouldGenerateTheStandardRoutes({ resource });

    it('should generate the publish route', () => {
      const { routes } = generateResource(resource);

      const { asFragment, getByText } = render(
        <Routes>{ routes() }</Routes>,
        {
          initialEntries: ['/rare-books/on-war/publish'],
          router: true,
          store: true,
        },
      );

      expect(getByText('Publish Page')).toBeVisible();

      expect(asFragment()).toMatchSnapshot();
    });

    it('should generate the published route', () => {
      const { routes } = generateResource(resource);

      const { asFragment, getByText } = render(
        <Routes>{ routes() }</Routes>,
        {
          initialEntries: ['/rare-books/published'],
          router: true,
          store: true,
        },
      );

      expect(getByText('Published Page')).toBeVisible();

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
