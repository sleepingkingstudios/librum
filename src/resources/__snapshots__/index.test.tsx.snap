// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resources generateResource() should generate the index route should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <h1>
      Rare Books
    </h1>
    <div
      class="min-h-[10rem] relative resources-table"
    >
      <span
        class="mock-data-table"
      >
        There are 3 books!
      </span>
    </div>
    <ul
      class="flex flex-col sm:flex-row page-breadcrumbs"
    >
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <span
          class="page-breadcrumb-empty"
        >
          Rare Books
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Resources generateResource() should generate the show route should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <h1>
      Gideon the Ninth
    </h1>
    <div
      class="min-h-[10rem] relative resource-block"
    >
      <span
        class="mock-data-table"
      >
        Showing book "Gideon the Ninth"
      </span>
    </div>
    <ul
      class="flex flex-col sm:flex-row page-breadcrumbs"
    >
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/rare-books"
        >
          Rare Books
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <span
          class="page-breadcrumb-empty"
        >
          Gideon the Ninth
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Resources generateResource() with extra pages should generate the index route should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <h1>
      Rare Books
    </h1>
    <div
      class="min-h-[10rem] relative resources-table"
    >
      <span
        class="mock-data-table"
      >
        There are 3 books!
      </span>
    </div>
    <ul
      class="flex flex-col sm:flex-row page-breadcrumbs"
    >
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <span
          class="page-breadcrumb-empty"
        >
          Rare Books
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Resources generateResource() with extra pages should generate the publish route 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <h1>
      Loading Rare Book
    </h1>
    <span>
      Publish Page
    </span>
    <ul
      class="flex flex-col sm:flex-row page-breadcrumbs"
    >
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/rare-books"
        >
          Rare Books
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <span
          class="page-breadcrumb-empty"
        >
          Loading
        </span>
      </li>
      <li
        class="page-breadcrumb"
      >
        <span
          class="page-breadcrumb-empty"
        >
          Publish
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Resources generateResource() with extra pages should generate the published route 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <h1>
      Published Rare Books
    </h1>
    <span>
      Published Page
    </span>
    <ul
      class="flex flex-col sm:flex-row page-breadcrumbs"
    >
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/rare-books"
        >
          Rare Books
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <span
          class="page-breadcrumb-empty"
        >
          Published
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Resources generateResource() with extra pages should generate the show route should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <h1>
      Gideon the Ninth
    </h1>
    <div
      class="min-h-[10rem] relative resource-block"
    >
      <span
        class="mock-data-table"
      >
        Showing book "Gideon the Ninth"
      </span>
    </div>
    <ul
      class="flex flex-col sm:flex-row page-breadcrumbs"
    >
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/rare-books"
        >
          Rare Books
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <span
          class="page-breadcrumb-empty"
        >
          Gideon the Ninth
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Resources generateResource() with scope: value should generate the index route should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <h1>
      Rare Books
    </h1>
    <div
      class="min-h-[10rem] relative resources-table"
    >
      <span
        class="mock-data-table"
      >
        There are 3 books!
      </span>
    </div>
    <ul
      class="flex flex-col sm:flex-row page-breadcrumbs"
    >
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/lending-library"
        >
          Lending Library
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <span
          class="page-breadcrumb-empty"
        >
          Rare Books
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Resources generateResource() with scope: value should generate the show route should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <h1>
      Gideon the Ninth
    </h1>
    <div
      class="min-h-[10rem] relative resource-block"
    >
      <span
        class="mock-data-table"
      >
        Showing book "Gideon the Ninth"
      </span>
    </div>
    <ul
      class="flex flex-col sm:flex-row page-breadcrumbs"
    >
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/lending-library"
        >
          Lending Library
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/lending-library/rare-books"
        >
          Rare Books
        </a>
      </li>
      <li
        class="page-breadcrumb"
      >
        <span
          class="page-breadcrumb-empty"
        >
          Gideon the Ninth
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
